File: src/main/java/gargoyle/calendar/core/CalCore.java
package gargoyle.calendar.core;

import gargoyle.calendar.util.Interpolation;
import gargoyle.calendar.util.Maps;
import gargoyle.calendar.util.resources.Resource;

import java.awt.*;
import java.awt.geom.Rectangle2D;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.time.*;
import java.time.format.TextStyle;
import java.time.temporal.ChronoField;
import java.time.temporal.TemporalAdjusters;
import java.time.temporal.TemporalField;
import java.time.temporal.ValueRange;
import java.util.Locale;
import java.util.Objects;


public final class CalCore {

    private static final String PARAM_YEAR = "year";
    private final CalConfig config;

    public CalCore(CalConfig config) {
        this.config = config;
    }

    private static void write(BufferedImage canvas, Rectangle2D area, String text, Font font, Color foreColor, Color backColor, boolean fill) {
        if (backColor != null) {
            Rectangle2D fillArea = fill || text == null ? area : CalUtil.metricText(canvas, Objects.requireNonNull(font), area, text);
            CalUtil.fill(canvas, backColor, fillArea);
        }
        if (text != null) {
            CalUtil.drawText(canvas, text, Objects.requireNonNull(foreColor), Objects.requireNonNull(font), area);
        }
    }

    private static long getLength(TemporalField field) {
        ValueRange range = field.range();
        return range.getMaximum() - range.getMinimum() + 1;
    }

    private static boolean isWeekend(DayOfWeek day) {
        return day == DayOfWeek.SUNDAY;
    }

    public void createWrite(Resource write, Year year, Locale locale, String formatName) throws IOException {
        BufferedImage image = createImage(year, locale);
        CalUtil.writeImage(image, write, formatName);
    }


    private BufferedImage createImage(Year year, Locale locale) throws IOException {
        int canvasWidth = config.getCanvasWidth();
        int canvasHeight = config.getCanvasHeight();
        BufferedImage canvas = CalUtil.createImage(canvasWidth, canvasHeight);
        printImage(canvas);
        printYearTitle(canvas, year);
        printYear(canvas, canvasHeight > canvasWidth, year, locale);
        return canvas;
    }

    private void printImage(BufferedImage canvas) throws IOException {
        int canvasWidth = config.getCanvasWidth();
        int canvasHeight = config.getCanvasHeight();
        BufferedImage original = CalUtil.readImage(config.getImageLocation());
        int originalWidth = original.getWidth();
        int originalHeight = original.getHeight();
        Rectangle2D inner = CalUtil.createRectangle(0, 0, originalWidth, originalHeight);
        Rectangle2D area = CalUtil.fit(CalUtil.createRectangle(0, 0, canvasWidth, canvasHeight), inner);
        CalUtil.fill(canvas, config.getBackgroundColor(), area);
        double areaWidth = area.getWidth();
        double areaHeight = area.getHeight();
        CalUtil.drawImage(canvas, area, CalUtil.scale(original, (int) areaWidth, (int) areaHeight));
    }

    private void printYearTitle(BufferedImage canvas, Year year) {
        int lineHeight = CalUtil.lineHeight(canvas, config.getDaysFont());
        int canvasWidth = canvas.getWidth();
        Rectangle2D area = CalUtil.createRectangle(0, 0, canvasWidth, lineHeight);
        String text = Interpolation.interpolate(config.getYearText(), Maps.of(PARAM_YEAR, year));
        write(canvas, area, text, config.getYearFont(), config.getYearForeColor(), config.getYearBackColor(), true);
    }

    private void printYear(BufferedImage canvas, boolean portrait, Year year, Locale locale) {
        long yearLength = getLength(ChronoField.MONTH_OF_YEAR);
        double sqrt = Math.sqrt(yearLength);
        int floor = (int) Math.floor(sqrt);
        int ceil = (int) Math.ceil(sqrt);
        int monthsPerRow = portrait ? floor : ceil;
        int monthsPerCol = portrait ? ceil : floor;
        double monthWidth = (double) canvas.getWidth() / monthsPerRow;
        double yearLineHeight = CalUtil.lineHeight(canvas, config.getYearFont());
        double monthHeight = (canvas.getHeight() - yearLineHeight) / monthsPerCol;
        double dayHeight = CalUtil.lineHeight(canvas, config.getDaysFont());
        for (Month month : Month.values()) {
            int monthIndex = month.ordinal();
            int col = monthIndex % monthsPerRow;
            int row = monthIndex / monthsPerRow;
            double x = col * monthWidth;
            double y = yearLineHeight + row * monthHeight;
            Rectangle2D area = CalUtil.createRectangle(x, y, monthWidth, monthHeight);
            YearMonth yearMonth = year.atMonth(month);
            printMonth(canvas, dayHeight, yearMonth, area, locale);
        }
    }

    private void printMonth(BufferedImage canvas, double dayHeight, YearMonth yearMonth, Rectangle2D monthArea, Locale locale) {
        int monthHeight = CalUtil.lineHeight(canvas, config.getMonthFont());
        double monthAreaX = monthArea.getX();
        double monthAreaY = monthArea.getY();
        double monthAreaWidth = monthArea.getWidth();
        Rectangle2D area = CalUtil.createRectangle(monthAreaX, monthAreaY, monthAreaWidth, monthHeight);
        Month month = yearMonth.getMonth();
        printMonthTitle(canvas, monthArea, month, locale);
        double dayWidth = area.getWidth() / getLength(ChronoField.DAY_OF_WEEK);
        double areaX = area.getX();
        double weekdaysY = area.getY() + monthHeight;
        printWeekdays(canvas, areaX, weekdaysY, dayWidth, dayHeight, locale);
        double daysY = area.getY() + monthHeight + dayHeight;
        printBackDays(canvas, areaX, daysY, dayWidth, dayHeight);
        printDays(canvas, yearMonth, areaX, daysY, dayWidth, dayHeight);
    }

    private void printMonthTitle(BufferedImage canvas, Rectangle2D monthArea, Month month, Locale locale) {
        String monthName = month.getDisplayName(TextStyle.FULL_STANDALONE, locale);
        int lineHeight = CalUtil.lineHeight(canvas, config.getDaysFont());
        double monthAreaX = monthArea.getX();
        double monthAreaY = monthArea.getY();
        double monthAreaWidth = monthArea.getWidth();
        Rectangle2D area = CalUtil.createRectangle(monthAreaX, monthAreaY, monthAreaWidth, lineHeight);
        write(canvas, area, monthName, config.getMonthFont(), config.getMonthForeColor(), config.getMonthBackColor(), true);
    }

    private void printWeekdays(BufferedImage canvas, double areaX, double areaY, double dayWidth, double dayHeight, Locale locale) {
        Color backColor = config.getWeekdaysBackColor();
        for (DayOfWeek dayOfWeek : DayOfWeek.values()) {
            int day = dayOfWeek.ordinal();
            double x = areaX + day * dayWidth;
            Rectangle2D dayArea = CalUtil.createRectangle(x, areaY, dayWidth, dayHeight);
            Color foreColor = isWeekend(dayOfWeek) ? config.getWeekdaysRedColor() : config.getWeekdaysForeColor();
            String weekdayName = dayOfWeek.getDisplayName(TextStyle.SHORT, locale);
            write(canvas, dayArea, weekdayName, config.getWeekdaysFont(), foreColor, backColor, true);
        }
    }

    private void printBackDays(BufferedImage canvas, double areaX, double areaY, double dayWidth, double dayHeight) {
        Color backColor = config.getDaysBackColor();
        long weekLength = getLength(ChronoField.DAY_OF_WEEK);
        long l = getLength(ChronoField.ALIGNED_WEEK_OF_MONTH) / weekLength;
        for (int row = 0; row < l; row++) {
            for (int col = 0; col < weekLength; col++) {
                double x = areaX + col * dayWidth;
                double y = areaY + row * dayHeight;
                Rectangle2D.Double dayArea = CalUtil.createRectangle(x, y, dayWidth, dayHeight);
                write(canvas, dayArea, null, null, null, backColor, false);
            }
        }
    }

    private void printDays(BufferedImage canvas, YearMonth yearMonth, double areaX, double areaY, double dayWidth, double dayHeight) {
        LocalDate firstDay = yearMonth.atEndOfMonth().with(TemporalAdjusters.firstDayOfMonth());
        DayOfWeek weekdayIndex = firstDay.getDayOfWeek();
        ValueRange range = firstDay.range(ChronoField.DAY_OF_MONTH);
        int weekLength = (int) getLength(ChronoField.DAY_OF_WEEK);
        int maximum = (int) range.getMaximum();
        int minimum = (int) range.getMinimum();
        for (int day = minimum; day <= maximum; day++) {
            long idx = weekdayIndex.ordinal() + day - 1;
            long col = idx % weekLength;
            long row = idx / weekLength;
            double x = areaX + col * dayWidth;
            double y = areaY + row * dayHeight;
            Color color = isWeekend(yearMonth.atDay(day).getDayOfWeek()) ? config.getDaysRedColor() : config.getDaysForeColor();
            Rectangle2D dayArea = CalUtil.createRectangle(x, y, dayWidth, dayHeight);
            write(canvas, dayArea, String.valueOf(day), config.getDaysFont(), color, null, true);
        }
    }

    @Override
    public String toString() {
        return String.format("CalCore{config=%s}", config);
    }
}

File: cal-core/src/main/kotlin/gargoyle/calendar/core/CalCore.kt
package gargoyle.calendar.core

import gargoyle.calendar.util.resources.Resource
import java.awt.Color
import java.awt.Font
import java.awt.geom.Rectangle2D
import java.awt.image.BufferedImage
import java.time.DayOfWeek
import java.time.Month
import java.time.Year
import java.time.YearMonth
import java.time.format.TextStyle
import java.time.temporal.ChronoField
import java.time.temporal.TemporalAdjusters
import java.time.temporal.TemporalField
import java.util.Locale
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.sqrt

class CalCore(private val config: CalConfig) {
    fun createWrite(write: Resource, year: Year, locale: Locale, formatName: String) {
        val image = createImage(year, locale)
        CalUtil.writeImage(image, write, formatName)
    }

    private fun createImage(year: Year, locale: Locale): BufferedImage {
        val canvasWidth = config.canvasWidth
        val canvasHeight = config.canvasHeight
        val canvas = CalUtil.createImage(canvasWidth, canvasHeight)
        printImage(canvas)
        printYearTitle(canvas, year)
        printYear(canvas, canvasHeight > canvasWidth, year, locale)
        return canvas
    }

    private fun printImage(canvas: BufferedImage) {
        val canvasWidth = config.canvasWidth
        val canvasHeight = config.canvasHeight
        val original = CalUtil.readImage(config.imageLocation)
        val originalWidth = original.width
        val originalHeight = original.height
        val inner: Rectangle2D = CalUtil.createRectangle(0.0, 0.0, originalWidth.toDouble(), originalHeight.toDouble())
        val area =
            CalUtil.fit(CalUtil.createRectangle(0.0, 0.0, canvasWidth.toDouble(), canvasHeight.toDouble()), inner)
        CalUtil.fill(canvas, config.backgroundColor, area)
        val areaWidth = area.width
        val areaHeight = area.height
        CalUtil.drawImage(canvas, area, CalUtil.scale(original, areaWidth.toInt(), areaHeight.toInt()))
    }

    private fun printYearTitle(canvas: BufferedImage, year: Year) {
        val lineHeight = CalUtil.lineHeight(canvas, config.daysFont)
        val canvasWidth = canvas.width
        val area: Rectangle2D = CalUtil.createRectangle(0.0, 0.0, canvasWidth.toDouble(), lineHeight.toDouble())
        val text = interpolate(config.yearText, mapOf(PARAM_YEAR to year))
        write(canvas, area, text, config.yearFont, config.yearForeColor, config.yearBackColor, true)
    }

    private fun interpolate(text: String, params: Map<String, Any>): String {
        var ret = text
        params.forEach { (key, value) ->
            ret = ret.replace("{$key}", value.toString())
        }
        return ret
    }

    private fun printYear(canvas: BufferedImage, portrait: Boolean, year: Year, locale: Locale) {
        val yearLength = getLength(ChronoField.MONTH_OF_YEAR)
        val sqrt = sqrt(yearLength.toDouble())
        val floor = floor(sqrt).toInt()
        val ceil = ceil(sqrt).toInt()
        val monthsPerRow = if (portrait) floor else ceil
        val monthsPerCol = if (portrait) ceil else floor
        val monthWidth = canvas.width.toDouble() / monthsPerRow
        val yearLineHeight = CalUtil.lineHeight(canvas, config.yearFont).toDouble()
        val monthHeight = (canvas.height - yearLineHeight) / monthsPerCol
        val dayHeight = CalUtil.lineHeight(canvas, config.daysFont).toDouble()
        for (month in Month.values()) {
            val monthIndex = month.ordinal
            val col = monthIndex % monthsPerRow
            val row = monthIndex / monthsPerRow
            val x = col * monthWidth
            val y = yearLineHeight + row * monthHeight
            val area: Rectangle2D = CalUtil.createRectangle(x, y, monthWidth, monthHeight)
            val yearMonth = year.atMonth(month)
            printMonth(canvas, dayHeight, yearMonth, area, locale)
        }
    }

    private fun printMonth(
        canvas: BufferedImage,
        dayHeight: Double,
        yearMonth: YearMonth,
        monthArea: Rectangle2D,
        locale: Locale
    ) {
        val monthHeight = CalUtil.lineHeight(canvas, config.monthFont)
        val monthAreaX = monthArea.x
        val monthAreaY = monthArea.y
        val monthAreaWidth = monthArea.width
        val area: Rectangle2D = CalUtil.createRectangle(monthAreaX, monthAreaY, monthAreaWidth, monthHeight.toDouble())
        val month = yearMonth.month
        printMonthTitle(canvas, monthArea, month, locale)
        val dayWidth = area.width / getLength(ChronoField.DAY_OF_WEEK)
        val areaX = area.x
        val weekdaysY = area.y + monthHeight
        printWeekdays(canvas, areaX, weekdaysY, dayWidth, dayHeight, locale)
        val daysY = area.y + monthHeight + dayHeight
        printBackDays(canvas, areaX, daysY, dayWidth, dayHeight)
        printDays(canvas, yearMonth, areaX, daysY, dayWidth, dayHeight)
    }

    private fun printMonthTitle(canvas: BufferedImage, monthArea: Rectangle2D, month: Month, locale: Locale) {
        val monthName = month.getDisplayName(TextStyle.FULL_STANDALONE, locale)
        val lineHeight = CalUtil.lineHeight(canvas, config.daysFont)
        val monthAreaX = monthArea.x
        val monthAreaY = monthArea.y
        val monthAreaWidth = monthArea.width
        val area: Rectangle2D = CalUtil.createRectangle(monthAreaX, monthAreaY, monthAreaWidth, lineHeight.toDouble())
        write(canvas, area, monthName, config.monthFont, config.monthForeColor, config.monthBackColor, true)
    }

    private fun printWeekdays(
        canvas: BufferedImage,
        areaX: Double,
        areaY: Double,
        dayWidth: Double,
        dayHeight: Double,
        locale: Locale
    ) {
        val backColor = config.weekdaysBackColor
        for (dayOfWeek in DayOfWeek.values()) {
            val day = dayOfWeek.ordinal
            val x = areaX + day * dayWidth
            val dayArea: Rectangle2D = CalUtil.createRectangle(x, areaY, dayWidth, dayHeight)
            val foreColor = if (isWeekend(dayOfWeek)) config.weekdaysRedColor else config.weekdaysForeColor
            val weekdayName = dayOfWeek.getDisplayName(TextStyle.SHORT, locale)
            write(canvas, dayArea, weekdayName, config.weekdaysFont, foreColor, backColor, true)
        }
    }

    private fun printBackDays(
        canvas: BufferedImage,
        areaX: Double,
        areaY: Double,
        dayWidth: Double,
        dayHeight: Double
    ) {
        val backColor = config.daysBackColor
        val weekLength = getLength(ChronoField.DAY_OF_WEEK)
        val l = getLength(ChronoField.ALIGNED_WEEK_OF_MONTH) / weekLength
        for (row in 0 until l) {
            for (col in 0 until weekLength) {
                val x = areaX + col * dayWidth
                val y = areaY + row * dayHeight
                val dayArea = CalUtil.createRectangle(x, y, dayWidth, dayHeight)
                write(canvas, dayArea, null, null, null, backColor, false)
            }
        }
    }

    private fun printDays(
        canvas: BufferedImage,
        yearMonth: YearMonth,
        areaX: Double,
        areaY: Double,
        dayWidth: Double,
        dayHeight: Double
    ) {
        val firstDay = yearMonth.atEndOfMonth().with(TemporalAdjusters.firstDayOfMonth())
        val weekdayIndex = firstDay.dayOfWeek
        val range = firstDay.range(ChronoField.DAY_OF_MONTH)
        val weekLength = getLength(ChronoField.DAY_OF_WEEK).toInt()
        val maximum = range.maximum.toInt()
        val minimum = range.minimum.toInt()
        for (day in minimum..maximum) {
            val idx = (weekdayIndex.ordinal + day - 1).toLong()
            val col = idx % weekLength
            val row = idx / weekLength
            val x = areaX + col * dayWidth
            val y = areaY + row * dayHeight
            val color = if (isWeekend(yearMonth.atDay(day).dayOfWeek)) config.daysRedColor else config.daysForeColor
            val dayArea: Rectangle2D = CalUtil.createRectangle(x, y, dayWidth, dayHeight)
            write(canvas, dayArea, day.toString(), config.daysFont, color, null, true)
        }
    }

    override fun toString(): String {
        return String.format("CalCore{config=%s}", config)
    }

    companion object {
        private const val PARAM_YEAR = "year"
        private fun write(
            canvas: BufferedImage,
            area: Rectangle2D,
            text: String?,
            font: Font?,
            foreColor: Color?,
            backColor: Color?,
            fill: Boolean
        ) {
            if (backColor != null) {
                val fillArea = if (fill || text == null) area else CalUtil.metricText(
                    canvas,
                    requireNotNull(font),
                    area,
                    text
                )
                CalUtil.fill(canvas, backColor, fillArea)
            }
            if (text != null) {
                CalUtil.drawText(canvas, text, requireNotNull(foreColor), requireNotNull(font), area)
            }
        }

        private fun getLength(field: TemporalField): Long {
            val range = field.range()
            return range.maximum - range.minimum + 1
        }

        private fun isWeekend(day: DayOfWeek): Boolean {
            return day == DayOfWeek.SUNDAY
        }
    }
}
